Do not allow closing workers to be awaited again

fjetter opened the pull request:
This fixes some of our deadlock situations causing tests to time out while closing a worker.

This is not a ready fix but rather a preliminary one to communicate the issue.

What's happening:

- A worker is closing
- Worker.status is set to `closing`. This will ensure `Worker.close` is idempotent and let every other caller to await finished, see https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/worker.py#L1647-L1648
- While it is closing, an incoming RPC triggers an [`await self`](https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/core.py#L465)
- The `await self` will attempt to start the worker again since `closing` is not part of `Status.ANY_RUNNING`
- The restart will set the status to `running` such that a second close attempt will actually try to run a close concurrently. the second call is actually what is deadlocking then while trying to shutdown the threadpool executor

Open questions:
- Is it justified that we have multiple close attempts?
- Why did this ever work?
- ...

cc @crusaderky @graingert 

closes https://github.com/dask/distributed/issues/5932

graingert commented on distributed/core.py:
```diff
@@ -280,7 +280,7 @@ def __await__(self):
             async with self._startup_lock:
-                if self.status in Status.ANY_RUNNING:
+                if self.status not in {Status.undefined, Status.created, Status.init}:
                     return self
```
start does nothing if the worker is `Status.closing`
https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/worker.py#L1522-L1527
then `__await__` immediately sets the worker to `Status.running` this is definitely not correct

graingert replied:
Nanny has an interesting behaviour with `Nanny.__await__` and `Nanny.start()`: https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/nanny.py#L341-L345

it sets its own `self.status = Status.running` or calls `self.close()`  - but if you `await Nanny(...)`  then whatever `Nanny.start()` does to `self.status` gets set back to `self.status = Status.running`

graingert replied:
this also interacts badly with `SpecCluster._correct_state_internal` https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/deploy/spec.py#L348-L352

`asyncio.wait(` doesn't propagate exceptions and so a worker can fail to start, and then be awaited a second time in `await w  # for tornado gen.coroutine support `

crusaderky commented on distributed/core.py:
```diff
@@ -280,7 +280,7 @@ def __await__(self):
             timeout = getattr(self, "death_timeout", 0)
             async with self._startup_lock:
-                if self.status in Status.ANY_RUNNING:
+                if self.status not in {Status.undefined, Status.created, Status.init}:
```
I'd rather have a new set in the Status enum, e.g. ``self.status not in Status.NEVER_STARTED``

fjetter replied:
as I mentioned above, this is not the final fix but rather an "FYI something here is broken". imo, the problem is a bit larger than just this particular set

fjetter commented on distributed/core.py:
```diff
@@ -247,7 +236,9 @@ def set_thread_ident():
         self.io_loop.add_callback(set_thread_ident)
-        self._startup_lock = asyncio.Lock()
+        self._started = asyncio.Event()
+        self.__status = Status.init
```
Reusing the `status` attribute for workers, nannies, schedulers and the core server is not easily possible without some major changes. There are many different semantics encoded and I don't believe the server base class should react to any status change of the child. I don't mind renaming this attribute but I believe there should be different stati

fjetter commented on distributed/core.py:
```diff
@@ -278,24 +273,19 @@ async def finished(self):
-                        self.status = Status.running
-                    except Exception:
-                        await self.close(timeout=1)
-                        raise TimeoutError(
-                            "{} failed to start in {} seconds".format(
-                                type(self).__name__, timeout
-                            )
-                        )
```
Raising this as a timeout error is just wrong

fjetter commented:
Working on this, I'm wondering if we should switch from an inheritance model to a composition model

graingert commented on distributed/core.py:
```diff
@@ -278,24 +273,19 @@ async def finished(self):
-                    await self.start()
-                    self.status = Status.running
+            if self.__status != Status.init:
+                await self._started.wait()
```
The worker could have previously failed to start and so this event will never be set

graingert replied:
I think the `_startup_lock` works better to make concurrent starts wait for the first start

graingert commented on distributed/utils_test.py:
```diff
@@ -1963,3 +1962,32 @@ def has_pytestmark(test_func: Callable, name: str) -> bool:
     return any(mark.name == name for mark in marks)
+
+
+E = TypeVar("E", bound=type)
```
```suggestion
E = TypeVar("E", bound=BaseException)
```

graingert commented on distributed/utils_test.py:
```diff
@@ -1963,3 +1962,32 @@ def has_pytestmark(test_func: Callable, name: str) -> bool:
     return any(mark.name == name for mark in marks)
+
+
+E = TypeVar("E", bound=type)
+
+
+@contextmanager
+def raises_with_cause(
+    expected_exception: E | tuple[E, ...],
+    match: str | None,
+    expected_cause: E | tuple[E, ...],
+    match_cause: str | None,
+) -> Generator[None, None, None]:
+    """Contextmanager to assert that a certain exception with cause was raised
+
+    Parameters
+    ----------
+    exc_type:
+    """
+    with pytest.raises(expected_exception, match=match) as exc_info:  # type: ignore
+        yield
```
pytest.raises expects a type[BaseException] not a type - so this should fix the `type: ignore`

the pytest stubs aren't installed in the mypy pre-commit hook - so you'll need to add that to additional_dependencies (imho we should run mypy in the conda environment and not with pre-commit - so all the stubs are always in place)

```suggestion
@contextmanager
def raises_with_cause(
    expected_exception: type[BaseException] | tuple[type[BaseException], ...],
    match: str | None,
    expected_cause: type[BaseException] | tuple[type[BaseException], ...],
    match_cause: str | None,
) -> Generator[None, object, None]:
    """Contextmanager to assert that a certain exception with cause was raised

    Parameters
    ----------
    exc_type:
    """
    with pytest.raises(expected_exception, match=match) as exc_info:
        yield
```

fjetter commented:
I chose a different path

1. No longer await the object itself. We don't want to restart, we merely want to check that it is already up. I added an event to track this
2. I still kept the refactorings about start I started with. To make things less entangled, I introduced `start_unsafe` which does not need to check for any concurrent startups, etc. and there is `Server.start` which wraps this and ensures that everything is handled properly.

mrocklin commented:
Checking in, what is the status here?

mrocklin commented:
I hope you don't mind, but I merged in main and pushed.  I'm hoping to trigger CI.

fjetter commented:
> Checking in, what is the status here?

waiting for CI. GH actions is in maintenance right now https://www.githubstatus.com/ I suppose this is why there are no builds scheduled

fjetter commented:
never mind, it's moving

fjetter commented:
`test_nanny_death_timeout` is failing all over the place due to a bug in `ConnectionPool`. The pool is improperly raising an `asyncio.CancelledError` as an `CommClosed`. I already started on a patch for this

graingert commented on distributed/tests/test_actor.py:
```diff
@@ -290,7 +290,6 @@ async def test_failed_worker(c, s, a, b):
 
     assert "actor" in str(info.value).lower()
```
how about moving this all into the pytest raises? `with pytest.raises(ValueError, match=r"Worker holding Actor was lost\.  Status: error'):`

graingert commented on distributed/utils_test.py:
```diff
@@ -2098,3 +2098,29 @@ def has_pytestmark(test_func: Callable, name: str) -> bool:
+
+    exc = exc_info.value
+    assert exc.__cause__
+    if not isinstance(exc.__cause__, expected_cause):
+        raise exc
```
```suggestion
    assert isinstance(exc.__cause__, expected_cause)
```

fjetter replied:
That's a slightly different thing, isn't it? To be honest, I'm not entirely sure which behavior is best. I suggest to keep it as is for now

jacobtomlinson commented:
@fjetter since this was merged we are seeing CI failures in `dask-kubernetes` when calling `client.wait_for_workers()`.

https://github.com/dask/dask-kubernetes/runs/6320080731?check_suite_focus=true

```python-traceback
______________________________ test_simplecluster ______________________________

k8s_cluster = <pytest_kind.cluster.KindCluster object at 0x7fb44ad9c070>
kopf_runner = <kopf._kits.runner.KopfRunner object at 0x7fb44afdf280>
gen_cluster = <function gen_cluster.<locals>.cm at 0x7fb44afb11f0>

    @pytest.mark.timeout(180)
    @pytest.mark.asyncio
    async def test_simplecluster(k8s_cluster, kopf_runner, gen_cluster):
        with kopf_runner as runner:
            async with gen_cluster() as cluster_name:
                scheduler_pod_name = "simple-cluster-scheduler"
                worker_pod_name = "simple-cluster-default-worker-group-worker"
                while scheduler_pod_name not in k8s_cluster.kubectl("get", "pods"):
                    await asyncio.sleep(0.1)
                while cluster_name not in k8s_cluster.kubectl("get", "svc"):
                    await asyncio.sleep(0.1)
                while worker_pod_name not in k8s_cluster.kubectl("get", "pods"):
                    await asyncio.sleep(0.1)
    
                with k8s_cluster.port_forward(f"service/{cluster_name}", 8786) as port:
                    async with Client(
                        f"tcp://localhost:{port}", asynchronous=True
                    ) as client:
>                       await client.wait_for_workers(2)

dask_kubernetes/operator/tests/test_operator.py:112: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/distributed/client.py:1328: in _wait_for_workers
    while n_workers and running_workers(info) < n_workers:
/opt/hostedtoolcache/Python/3.9.12/x64/lib/python3.9/site-packages/distributed/client.py:1321: in running_workers
    [
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <dict_valueiterator object at 0x7fb44886e220>

        [
            ws
            for ws in info["workers"].values()
>           if ws["status"] == Status.running.name
        ]
    )
E   KeyError: 'status'
```
